File: 2024/Day01.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>

#include "utils.hpp"

#include "Days.hpp"

Day01::Day01(string path) : AdventDay(path) {
    day = 1;
    title = "Historian Hysteria";
}

void Day01::part1(vector<string> lines) {
  vector<long> left, right;

  for(const string& line : lines) {
      vector<long> lr = extractLongs(line);
      left.push_back(lr[0]);
      right.push_back(lr[1]);
  }

  sort(left.begin(), left.end());
  sort(right.begin(), right.end());

  long res = 0;
  for(unsigned int i = 0; i < left.size(); i++) {
    res += abs(left[i] - right[i]);
  }

  printLongResult(res);
}

void Day01::part2(vector<string> lines) {
  vector<long> left;
  map<long, long> rightCount;

  for(const string& line : lines) {
      vector<long> lr = extractLongs(line);

    left.push_back(lr[0]);

    long right = lr[1];

    if(rightCount.contains(right)) {
      rightCount[right]++;
    } else {
      rightCount[right] = 1;
    }
  }

  long res = 0;
  for(const long& l : left) {
    if(rightCount.contains(l)) {
      res += l * rightCount[l];
    }
  }

  printLongResult(res);
}

File: 2024/python/day01.py
from aoc import *

def part1(lines):
    left = []
    right = []
    for l in lines:
        (a,b) = [int(x) for x in l.split()]
        left.append(a)
        right.append(b)

    left.sort()
    right.sort()
    return sum(abs(left[i]-right[i]) for i in range(len(left)))

def part2(lines):
    left = []
    right = {}
    for l in lines:
        (a,b) = [int(x) for x in l.split()]
        left.append(a)
        right[b] = right.get(b, 0) + 1
    return sum([right.get(x,0)*x for x in left])

AoCRunnerAll(1, 'Historian Hysteria', part1, part2)
